general/testing.md

```diff
@@ -1,5 +1,19 @@
 # Recommendations for testing
 
+## Testing UI components
+
+-   Snapshots SHOULD only be used as "smoke tests": they are very good for verifying that a component renders at all, but they are _not_ good for communicating intent or what is important. For example, when something in a large snapshot changes, there is nothing intrinsic in the snapshot to communicate whether any particular part of it is important.
+-   Shipping test attributes to the client is overtly bad because it increases network payloads, parse times, DOM bloat etc, and ends up becoming an unofficial API.
+    -   Therefore, we MUST NOT send "data-testid" to the client.
+-   The code and markup that is exercised in the test environment should match that which is used in the production environment, otherwise you are not actually testing the thing that you would like to test.
+    -   Because of the recommendation above, any usage of "data-testid" MUST be stripped from when outside the test environment.
+    -   But because of the desire to not have divergent code paths, "data-testid" SHOULD NOT be used, except as an interim measure.
+-   In the absence of "data-testid", we seek the most stable way to target elements in tests ("stable" in the sense that tests are less likely to break as a result of superficial changes); we think the methods that are most likely to remain stable are the following and therefore recommend that we SHOULD use:
+    -   Attributes or content that are significant to the user (for example, the text of a button such as "Publish"); at the end of the day, the visible (or audible) content and behavior of a product is literally the reason we build products, so the highest value tests should focus on aspects that are perceptible and observable to the user. In the button example, if the text changes, then we _want_ the test to break.
```

Comment:
How does that deal with localization?

Reply:
I think that in the same way you need to write tests that don't produce different output (and break) when run in different time zones, we need to isolate our test environment from locale differences. That is, I expect almost 100% of test runs to be locked to a specific localization, and then have a small number of instances where we effectively "sample" other localizations to make sure that in principal, things still work.

Reply:
Yeah, you're probably right... we don't really have a way to serve the right translations anyways...

One thing that @bfarache brought up, though, was how could we test that different translations reflected properly in the UI (german for example being usually longer than english). I'm thinking this should probably be tested somewhere else, just to double-check.

Reply:
> I'm thinking this should probably be tested somewhere else

Yeah, I don't know if we have any tooling that could help us check for catastrophic layout failures (or even non-catastrophic but still unpleasantly usability problems) in an automated way. It seems like it would be pretty brittle even if we had it (how would you distinguish between "broken" layout and merely "different" layout?). Maybe we already do, but part of the manual QA for every new feature could be to test in English, German (because of the long words), and probably a RTL language as well.

And I think we'd want some non-visual tests that run in a language other than English just to minimally exercise the localization machinery (or part of it, because as you said in another thread, the `Liferay.Language` stuff is hard to test end-to-end because it spans so many levels).